app/src/main/java/com/javahelp/frontend/activity/ProviderRegistrationActivity.java

```diff
@@ -25,14 +50,109 @@ protected void onCreate(Bundle savedInstanceState) {
+     * Called when the Register button is clicked
+     *
+     * @param v {@link View} that was clicked
+     */
+    private void registerClick(View v) {
+        providerRegistrationViewmodel.setUsername(binding.username.getText().toString());
+        providerRegistrationViewmodel.setPassword1(binding.password.getText().toString());
+        providerRegistrationViewmodel.setPassword2(binding.repassword.getText().toString());
+        providerRegistrationViewmodel.setProviderUserInfo(binding.email.getText().toString(),
+                binding.firstname.getText().toString(),
+                binding.lastname.getText().toString(),
+                binding.home.getText().toString(),
+                binding.phonenumber.getText().toString());
+        registerAttempt();
+    }
+
+    /**
+     * Initiates a register request
+     */
+    private void registerAttempt() {
+        if (binding.username.getText().toString().isEmpty() || binding.password.getText().toString().isEmpty()
+        || binding.repassword.getText().toString().isEmpty()) {
+            providerRegistrationViewmodel.setRegisterError("Please enter username and password");
+            return; // early return on invalid username or password
+        }else if(!binding.password.getText().toString().equals(binding.repassword.getText().toString())){
+            providerRegistrationViewmodel.setRegisterError("Passwords do not match");
+            return;
+            }else if(binding.email.getText().toString().isEmpty() || binding.firstname.getText().toString().isEmpty()
+        || binding.lastname.getText().toString().isEmpty() || binding.home.getText().toString().isEmpty()
+        ||binding.phonenumber.getText().toString().isEmpty()){
+            providerRegistrationViewmodel.setRegisterError("Please enter your personal information");
+            return;
+        }
+    }
+
+    /**
+     * Updates this {@link RegisterResult} to reflect the passed register result
+     *
+     * @param registerResult {@link Optional} {@link RegisterResult} to update on
+     */
+    private void updateOnRegisterResult(Optional<RegisterResult> registerResult) {
+        if(registerResult.isPresent()){
+            updateOnPresentRegisterResult(registerResult.get());
+        } else {
+            binding.registerErrorText.setVisibility(View.GONE);
+        }
+    }
+
+    /**
+     * Updates the UI based on a {@link RegisterResult}. This is called with non-null register results, meaning
+     * a registration has been attempted, and has either succeeded or failed. This is unlike updateOnRegisterResult
+     * which is called with an {@link Optional} {@link RegisterResult}, and calls this method if that
+     * {@link Optional} is present.
+     *
+     * @param result {@link RegisterResult} of login attempt
+     */
+    private void updateOnPresentRegisterResult(RegisterResult result) {
+        if (result.isSuccess()) {
+            binding.registerErrorText.setText("Register successful");
```

Comment:
Based on your Javadoc description, I personally think `binding.registerErrorText` could be renamed to `bingding.registerResultText` to avoid misunderstanding.